### ISSUE TITLE
lizardfs metadata performance becomes slow

### ISSUE THREAD

**Issue by nadiasvertex** [lizardfs/lizardfs#559]
I have a small, 3-node LFS v3.11.0 cluster serving another 4 nodes with a fairly high duty cycle. It has been working great until just recently. I didn't make any configuration changes to the system.

Now the performance of something "ls -l" becomes very poor. For example, on a folder with just 5 sub folders it may take a second to run. On a folder with tens of files it takes minutes. Interestingly, if I restart any chunkserver on any of the storage nodes, performance improves dramatically for a while. Eventually it slows down again.

The machines it is running on have two cores / 4GB, and mfsmaster is using 100% of one core all the time.  mfsmaster is using about 10% of the memory on the system, or ~400Mb. The system shows 1GB used of 4GB, so it doesn't seem like resource exhaustion. 

Any diagnostic suggestions or other information would be great.

**Comment by Blackpaw** [comment]
Do you have undergoal chunks? there is a bug in3.11.0 (https://github.com/lizardfs/lizardfs/issues/555) that causes mfsmaster to hog the cpu for an hour or more after healing chunks.

**Comment by nadiasvertex** [comment]
I don't now, but I did. Is there a way to kick it and make it stop?

**Comment by Blackpaw** [comment]
Oops, better discussion of it here:  https://github.com/lizardfs/lizardfs/issues/547

Maybe if you set OPERATIONS_DELAY_DISCONNECT to 30 then restart msfmaster? that risks more undergoal chunks of course.

It should return to normal within an hour.

**Comment by nadiasvertex** [comment]
I'll look at that. The odd thing is that, if I restart any chunkserver,
everything gets fast again for a while. Eventually it slows down again.

**Comment by Blackpaw** [comment]
POOMA U of course but I think that while its processing chunks the master is waiting on io and not spinning its wheels - once there are no chunk operations it goes into a loop for OPERATIONS_DELAY_DISCONNECT  seconds.

Actually if you:
* set OPERATIONS_DELAY_DISCONNECT =30 (or some small figure)
* reload the mfsmaster (not restart)
* restart a chunkserver

Then it should heal the chunks, spin its wheels for 30 seconds, then be ok.

**Comment by nadiasvertex** [comment]
I tried that, but it doesn't make a difference to the master. It continues to use 100% CPU. I also see that operations are FAST right after I restart a chunk server, but get SLOW over time. The healing happens very fast, but eventually metatdata operations become unreasonably slow.

**Comment by nadiasvertex** [comment]
I appreciate your help. Just a few more data points, the machines running mfsmaster (either as master or as shadow) have been pegging their CPUs since I upgraded. I set the OPERATIONS_ values to small numbers (10 and 30). I have reloaded and restarted the master mfsmaster, also restarting the chunkserver on the same machine.

Some high level of activity seems to be causing the entire cluster to slowly degrade performance over time. If I leave it overnight, when I come in the next morning everything is running so slowly it appears to be hung. It is making progress, however. If I leave them alone they will eventually finish.

**Comment by Blackpaw** [comment]
That is concerning

I only run a test VM cluster myself, so not so affected by it. But what I did di was install the packag src, apply the patch and rebuild/install it to test the patch, which did work. This was on debian jessie.

You could do that as well or I could send you the packages.

**Comment by DarkHaze** [comment]
Please give us more information about your setup:

- system version
- number of files in your folder (only 5 subfolders ?)
- screenshot from the cgi (info + servers)

**Comment by nadiasvertex** [comment]
That particular folder only had 5 subfolders, yes. No files. There are, of course, many more files in the whole system.

$ lizardfs-admin info 127.0.0.1 9421
LizardFS v3.11.0
Memory usage:   418MiB
Total space:    936GiB
Available space:        397GiB
Trash space:    104KiB
Trash files:    488
Reserved space: 0B
Reserved files: 0
FS objects:     946507
Directories:    67601
Files:  878906
Chunks: 754724
Chunk copies:   1093630
Regular copies (deprecated):    1093630

$ lizardfs-admin chunks-health 127.0.0.1 9421
Chunks availability state:
        Goal    Safe    Unsafe  Lost
        1       -       414829  1
        2       338895  -       -

Chunks replication state:
        Goal    0       1       2       3       4       5       6       7       8       9       10+
        1       414829  1       -       -       -       -       -       -       -       -       -
        2       338895  -       -       -       -       -       -       -       -       -       -

Chunks deletion state:
        Goal    0       1       2       3       4       5       6       7       8       9       10+
        1       414830  -       -       -       -       -       -       -       -       -       -
        2       338888  7       -       -       -       -       -       -       -       -       -

$ lizardfs-admin list-chunkservers localhost 9421
Server 10.116.23.147:9422:
        version: 3.11.0
        label: san
        chunks: 363k
        used space: 186GiB / 318GiB
        chunks marked for removal: 0
        used space marked for removal: 0B / 0B
        errors: 0
Server 10.116.23.143:9422:
        version: 3.11.0
        label: san
        chunks: 360k
        used space: 175GiB / 300GiB
        chunks marked for removal: 0
        used space marked for removal: 0B / 0B
        errors: 0
Server 10.116.23.148:9422:
        version: 3.11.0
        label: san
        chunks: 371k
        used space: 178GiB / 318GiB
        chunks marked for removal: 0
        used space marked for removal: 0B / 0B
        errors: 0

FAST TIMING
--------------------

$ time ls
catalog  config  data  jobs  snapshots

real    0m0.003s
user    0m0.000s
sys     0m0.000s

$ ping usmeps003.bethel.jw.org
PING usmeps003 (10.116.23.143) 56(84) bytes of data.
64 bytes from usmeps003 (10.116.23.143): icmp_seq=1 ttl=64 time=0.156 ms
64 bytes from usmeps003 (10.116.23.143): icmp_seq=2 ttl=64 time=0.127 ms
64 bytes from usmeps003 (10.116.23.143): icmp_seq=3 ttl=64 time=0.143 ms
64 bytes from usmeps003 (10.116.23.143): icmp_seq=4 ttl=64 time=0.169 ms


SLOW TIMING
--------------------
(This is coming, I just reset the system to get fast timings.)

**Comment by nadiasvertex** [comment]
Another piece of data, if I restart a chunkserver, the system stays fast for whatever time OPERATIONS_DELAY_DISCONNECT is set to. Then the slowdown documented above occurs.

**Comment by gutleib** [comment]
I can fully confirm this bug.
After about a day of usage:
```
real    0m0.613s
user    0m0.000s
sys     0m0.004s
```
After resetting a chunkserver:
```
real    0m0.006s
user    0m0.000s
sys     0m0.000s
```
Also, if there is no fix so far, we need a way to downgrade.

A most logical way - to get less recent version of mfsmaster (e.g. 3.10.4) connect to cluster of 3.11.0, but now it is blocked with `mfsmaster[8320]: MLTOMA_REGISTER_SHADOW - rejected old client (v3.10.` message.

I guess, I'll make another ticket for downgrade tool.

**Comment by cypromis** [comment]
Have you tried 3.11.1, does it fix your issue ?

**Comment by gutleib** [comment]
We got patch yesterday, but were waiting to confirm mid-term stability.

Yes, 3.11.1 fixes this, we are 15+ hours not observing metadata slowness.

**Comment by nadiasvertex** [comment]
I took the patch this morning, but mfsmaster had trouble starting up. It failed the first time, then it took it about half an hour to finish it's startup process. However, it did initialize correctly after that. I will report back on performance after it has run for a few hours.

Thanks!

**Comment by nadiasvertex** [comment]
So far performance is remained good on this cluster! I'll continue monitoring and then close this bug if things stay good overnight.

**Comment by nadiasvertex** [comment]
Performance has stayed high. Looks like the 3.11.1 patch fixed it. Thanks!
